bpf/bpf_host.c

```diff
@@ -1292,26 +1292,61 @@ int cil_from_host(struct __ctx_buff *ctx)
 	return handle_netdev(ctx, true);
 }
 
+#if defined(ENABLE_IPSEC) && defined(ENABLE_IPV4) && defined(TUNNEL_MODE)
+/*
+ * If the traffic is indeed overlay traffic and it should be encrypted
+ * CTX_ACT_REDIRECT is returned, unless an error occurred, and the caller can
+ * return this code to TC.
+ *
+ * CTX_ACT_OK is returned if the traffic should continue normal processing.
+ *
+ * IS_ERR can be used to determine if this function ran into an error.
+ */
+static __always_inline int do_encrypt_overlay(struct __ctx_buff *ctx,
+					      __u16 proto)
+{
+	int ret = CTX_ACT_OK;
+	struct iphdr __maybe_unused *ipv4;
+	void __maybe_unused *data, *data_end = NULL;
+
+	/* only support vxlan for now */
+	if (TUNNEL_PROTOCOL != TUNNEL_PROTOCOL_VXLAN)
+		return ret;
+
+	if (proto != bpf_htons(ETH_P_IP))
+		return ret;
+
+	if (!revalidate_data(ctx, &data, &data_end, &ipv4))
+		return DROP_INVALID;
+
+	if (!tunnel_skb_is_vxlan_v4(data, data_end, ipv4, TUNNEL_PORT))
+		return ret;
+
+	if (tunnel_vxlan_get_vni(data, data_end, ipv4) ==
+				 MARK_MAGIC_ENCRYPT) {
```

Comment:
Solid point. 

The reason I opted to check check explicitly the encrypt bits is because we are checking the mark in a new code path. I was not sure if say, this code path did not expect to ever see the encrypt bits set, and by **chance** a different mark comes in with the encrypt bits set (along with others) but without the intention for encryption (we use marks alot lol).

That being said, if we can work out that this 'over matching' of the mark cannot occur or simply won't, this can change to a '&' where we just mask-check the encryption bits.

Reply:
Ugh, I would hate to cripple cluster mesh support for this right off the bat. 

But let's explore that path a bit.
When cluster mesh is disabled, do we still retain the 8 bits in the security ID as reserved? I see a bunch of code like this:
```
static __always_inline __u32
extract_cluster_id_from_identity(__u32 identity)
{
	return (__u32)(identity >> IDENTITY_LEN);
}
``` 
Making me feel like like the ratio of identity to cluser_id bits is variable via config? Not sure if this means that when cluster mesh is disabled, sec_id's become a full 32bits?

Reply:
> The alternative is to get the inner destination IP address and use it to lookup the SPI in the ipcache.

^ This seems viable.

If we combine this with the following then I think we can mitigate some of the impacts:
1. Use a reserved identity (for example, `11`) and not `MARK_MAGIC_ENCRYPT` / `0x0E00`, which is a cluster-scoped identity, then we can avoid overloading that security identity value. We could guarantee that we never expect VNI `11` to be legitimately used as a security identity for policy enforcement purposes.
2. Ensure the datapath replaces this reserved identity with the real identity, and drop traffic that attempts to leave the node with this reserved identity.

Reply:
I want to confirm the approach I'm thinking here, since Paul's exact words:

> The alternative is to get the inner destination IP address and use it to lookup the SPI in the ipcache.

Are not the same as the solution we were discussion. 

When we get to `do_encrypt_overlay` function in this PR and after we confirm that `vxlan.vni == 11` we can replace the VNI. 

At this point, we will sniff into the vxlan packet to identify the **source** address of the encapsulated packet, perform a local endpoint map lookup for the source pod, and rewrite the VNI field with the discovered security ID, set the encryption **mark** and then redirect it. 

I'm not sure if you have yet another idea here Paul involving grabbing the SPI by sniffing at the destination of the encap'd packet?

Reply:
To get the SPI we need to lookup the inner destination IP, but we can probably get the SPI via the updated node ID map as mentioned in the other thread. To get the source security identity, you need to lookup the inner source IP. So Joe's solution makes sense to me here.

Note that the second approach I proposed, which avoids parsing VXLAN and doing an ipcache lookup, is incompatible with Cilium Mesh, not with clustermesh.

Reply:
> To get the SPI we need to lookup the inner destination IP, but we can probably get the SPI via the updated node ID map as mentioned in the other thread.
You do mean a valid point here. 
Since we are now heading towards sniffing into the vxlan packet, we could just feed the dest to find the SPI as well... Which would mean less code changes :thinking: 

> So Joe's solution makes sense to me here.
:+1:

> Note that the second approach I proposed, which avoids parsing VXLAN and doing an ipcache lookup, is incompatible with Cilium Mesh, not with clustermesh.
Ugh sorry, I confused my meshes :(

Reply:
> Since we are now heading towards sniffing into the vxlan packet, we could just feed the dest to find the SPI as well... Which would mean less code changes and an alternative to placing it in NodeID map 🤔

Yep. It's one additional LPM lookup vs. a hashmap lookup, but I'm not sure it makes a big difference considering all of this is encrypted anyway.

Reply:
After some more thought Paul, we STILL may have situations where we are trying to get some traffic not related to a destination security ID over to the other node. 

Therefore, we still need to do SPI lookup via the InternalNodeIP or CiliumNodeIP. It probably makes more sense to just do this in the NodeID map after all...

Reply:
This should be fixed with latest push. Unresolve if you feel otherwise